Add support for ADAQ42xx

machschmitt opened the pull request:
# Description

AD4630 and ADAQ42xx designs are very similar, differing mainly on the availability (or not) of PGIA gain.
Add input scale properties to handle ADAQ42xx PGIA control and add ADAQ4216, ADAQ4220, and ADAQ4224 to the list of supported devices.

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How has this been tested?

This has been tested on ZebBoard with EVAL-ADAQ4224-FMCZ (setup by @ladace).

**Test Configuration**:
* Hardware: ZebBoard with EVAL-ADAQ4224-FMCZ 
* OS: Very likely Kuiper Linux

From iio_info:
```
	hdl_system_id: [AD463XADAQ42XXN0_CLKMODE0_NUMOFSDI1_CAPTUREZONE2_DDREN0/ad463x_adaq42xx_zed] [sys rom custom string placeholder] on [zed] git branch [dev_adaq4224_final] git [7df56bd2d04bdccc44d9bbfa94c800b828abb80f] dirty [2023-11-01 14:03:21] UTC
	hw_model: EVAL-ADAQ4224-FMCZ on Xilinx Zynq ZED
```
- [x] Test A: Ran the newly provided example for adaq4224 remotely with
`python3 examples/ad4630/ad4630_example_scale.py "ip:<address_omitted>" "adaq4224"` 
- [x] Test B: Ran the newly provided scale attribute test with
`pytest -vs --emu --custom-hw-map=test/emu/hardware_map.yml -k 'not prod' test/test_ad4630.py`


# Documentation

Linux kernel support has been included in PR [#2262](https://github.com/analogdevicesinc/linux/pull/2262/).

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have signed off all commits and they contain "Signed-off by: <insert name>"
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

tfcollins commented on adi/ad4630.py:
```diff
@@ -166,3 +170,70 @@ def calibscale(self):
+
+    def __init__(self, uri="", device_name="adaq4224"):
+
+        context_manager.__init__(self, uri, self._device_name)
+
+        compatible_parts = ["adaq4224", "adaq4216", "adaq4220"]
+
+        if device_name not in compatible_parts:
+            raise Exception(
+                "Not a compatible device: "
+                + str(device_name)
+                + ". Please select from "
+                + str(compatible_parts)
+            )
+        else:
+            self._ctrl = self._ctx.find_device(device_name)
+            self._rxadc = self._ctx.find_device(device_name)
+
+        self.__init_channels__(self)
```
This constructor is almost identical to the parent. Can you move "compatible_parts" to a class prop and set it uniquely for parent and child

machschmitt replied:
Sure, made the child class use parent's constructor on V2.

tfcollins commented on adi/ad4630.py:
```diff
@@ -166,3 +170,70 @@ def calibscale(self):
+            """Set calibration scale value."""
+            self._set_iio_attr(self.name, "calibscale", False, calibscale, self._ctrl)
+
+        @property
+        def scale_available(self):
+            """Provides all available scale(gain) settings for the ADAQ42xx channel"""
+            return self._get_iio_attr(self.channel[0].name, "scale_available", False)
+
+        @property
+        def scale(self):
+            """ADAQ42xx channel scale"""
+            return float(self._get_iio_attr_str(self.name, "scale", False))
+
+        @scale.setter
+        def scale(self, value):
+            self._set_iio_attr(self.name, "scale", False, str(Decimal(value).real))
```
These are the only new ones. Can you inherit from parent the reused methods to reduce duplication as well. Do you know if these exist in the parent driver but just not implemented in that class?

machschmitt replied:
No, the scale_available attribute is not available for ad4630/ad4030 devices.
adaq4224 has a PGA gain which allows the user to scale the input voltage to the ADC.
To support that, adaq4224 has the scale_available attribute and also allows setting the current scale through the scale attribute.
Made child's _channel inherit from parent class so it reuses the other methods.

tfcollins reviewed (changes requested):
We can get a bit better reuse on the parent class existing methods

machschmitt commented:
Thank you for your review @tfcollins.

I made the changes you asked for.

Change log V1 -> V2:
- Made compatible_parts a class attribute and used it to simplify adaq4224 subclass constructor.
- Made adaq4224 class _channel inherit from ad4630 _channel so it reuses the methods from the parent class.

tfcollins commented on adi/ad4630.py:
```diff
@@ -25,25 +26,27 @@ class ad4630(rx, context_manager, attribute):
-
-        if device_name not in compatible_parts:
+        if device_name not in self.__compat_parts__(self):
             raise Exception(
                 "Not a compatible device: "
                 + str(device_name)
                 + ". Please select from "
-                + str(compatible_parts)
+                + str(self.__compat_parts__(self))
```
Just use the class property  _compatible_parts directly here. There is no point making an extra method here.

machschmitt replied:
If I replace `self.__compat_parts__(self)` by `_compatible_parts` then `pre-commit run --all-files` says `_compatible_parts` is undefined. I could try `ad4630._compatible_parts` but then I presume the call from the child class would go wrong.

tfcollins commented on adi/ad4630.py:
```diff
@@ -166,3 +169,33 @@ def calibscale(self):
+        return adaq42xx._compatible_parts
+
+    def __init__(self, uri="", device_name="adaq4224"):
+        super().__init__()
```
need to pass uri and device_name to super

machschmitt replied:
done

tfcollins commented on adi/ad4630.py:
```diff
@@ -166,3 +169,33 @@ def calibscale(self):
+
+    _compatible_parts = ["adaq4224", "adaq4216", "adaq4220"]
+
+    def __compat_parts__(self):
+        return adaq42xx._compatible_parts
```
Remove

tfcollins reviewed (commented):
- Add the necessary import in __init__.py
- Include a new context file for the new system
- Add a new test since the driver has more properties

machschmitt commented:
I don't have adqa4224 but do have ad4030. I was trying to set up ad4030 with zed board to test this (even if indirectly), but something didn't work on the hdl/linux side. Will come back to this task when able to test the changes. By the way, should this PR also include a device XML file generated with `xml_gen`?

Change log V2 -> V3:
- Made child constructor pass uri and device_name to super

tfcollins commented:
> I don't have adqa4224 but do have ad4030. I was trying to set up ad4030 with zed board to test this (even if indirectly), but something didn't work on the hdl/linux side. Will come back to this task when able to test the changes. By the way, should this PR also include a device XML file generated with `xml_gen`?

Yes for the new component

machschmitt commented:
Hi @tfcollins, 

After testing this on ZebBoard with EVAL-ADAQ4224-FMCZ it was possible to make the code cleaner.
I updated the PR description with more detailed info about the tests.
Also, the related Linux kernel changes have been merged.
It is easier to squash than to split so I made the changes to attribute tests as a group of three small patches.
Do I add adaq4224 to tasks.py ignore list to pass checkparts?

Change log V3 -> V4
- Simplified back ad4630/adaq4224 constructors
- Fixed scale_available property which was broken
- Fixed bug on ad4630_example_simple_plot.py
- Added simple adaq4224 example featuring adjustable input scale.
- Minor neats to attribute tests to enable testing multiple values of channel attributes
- Added adaq4224 xml file
- Added test for new scale attribute
